dell/csi-baremetal
# [ISSUE-849] Call vgs refresh after each lvg-disk reinsert

## Purpose
### Resolves #849 

- Refactor `VolumeGroupScan...` events. On `checkVGErrors` call:
1. VolumeGroupScanInvolved (ever)
2. VolumeGroupScanFailed/VolumeGroupScanNoErrors/VolumeGroupScanErrorsFound (based on a result)
- Add checking all lvs via lsblk for `checkVGErrors`
- Refactor action after lvg disk OFFLINE->ONLINE:
1. `vgchange --refresh`
2. check io errors

## PR checklist

## Testing
1, Create sts with LVG volumes
```
kubectl apply -f tests/app/nginx-lvg.yaml
```
2. Find node, csibmnode and drive related to one of the volumes
```
hostname1:~ # kubectl get drive -o wide ba6533a0-4c3f-4214-b708-fb5704bbf826 

NAME                                   SIZE            TYPE   HEALTH   STATUS   USAGE    SYSTEM   PATH       SERIAL NUMBER   NODE                                   SLOT

ba6533a0-4c3f-4214-b708-fb5704bbf826   8001000000000   HDD    GOOD     ONLINE   IN_USE            /dev/sdl   VDH1AAND        ebcb42dd-656b-4b37-adb8-b47fe50c3a93   22
```
3. On a node with drive: mask drive via SCSI settings
```
echo 1 > /sys/class/scsi_device/0\:0\:11\:0/device/delete
```
4. Wait drive is in OFFLINE state
```
hostname1:~ # kubectl get drive -o wide ba6533a0-4c3f-4214-b708-fb5704bbf826 

NAME                                   SIZE            TYPE   HEALTH    STATUS    USAGE    SYSTEM   PATH       SERIAL NUMBER   NODE                                   SLOT

ba6533a0-4c3f-4214-b708-fb5704bbf826   8001000000000   HDD    UNKNOWN   OFFLINE   IN_USE            /dev/sdl   VDH1AAND        ebcb42dd-656b-4b37-adb8-b47fe50c3a93   22
```
5. Return drive
```
echo "- - -" > /sys/class/scsi_host/host0/scan
```
6. Wait drive is in ONLINE state
```
hostname1:~ # kubectl get drive -o wide ba6533a0-4c3f-4214-b708-fb5704bbf826 

NAME                                   SIZE            TYPE   HEALTH   STATUS   USAGE    SYSTEM   PATH       SERIAL NUMBER   NODE                                   SLOT

ba6533a0-4c3f-4214-b708-fb5704bbf826   8001000000000   HDD    GOOD     ONLINE   IN_USE            /dev/sdl   VDH1AAND        ebcb42dd-656b-4b37-adb8-b47fe50c3a93   22
```
7. Check LVG events
```
Events:
  Type     Reason                         Age   From                Message
  ----     ------                         ----  ----                -------
  Warning  VolumeGroupReactivateInvolved  105s  csi-baremetal-node  Trying to re-activate volume group
  Normal   VolumeGroupScanInvolved        105s  csi-baremetal-node  Check for IO errors
  Normal   VolumeGroupScanNoErrors        105s  csi-baremetal-node  No errors was found
```
8. Check node logs
```
May 17 14:33:01.670 [ERRO] [VolumeManager] [reactivateVG] [26576579-eb79-4465-af76-b170909b1b14] [pkg/node/volumemgr.go:1080] [reactivateVG] IO errors detected for volume group 26576579-eb79-4465-af76-b170909b1b14

May 17 14:33:01.670 [INFO] [LVM] Trying to re-activate volume group 26576579-eb79-4465-af76-b170909b1b14

May 17 14:33:01.878 [DEBU] [Executor] [/sbin/lvm vgchange --refresh 26576579-eb79-4465-af76-b170909b1b14] [208.351545ms] [208351545] stdout: 

May 17 14:33:01.878 [INFO] [VolumeManager] [checkVGErrors] [26576579-eb79-4465-af76-b170909b1b14] Scan volume group 26576579-eb79-4465-af76-b170909b1b14 for IO errors

May 17 14:33:01.986 [DEBU] [Executor] [/sbin/lvm vgscan] [107.581582ms] [107581582] stdout:   Found volume group "26576579-eb79-4465-af76-b170909b1b14" using metadata type lvm2

  Found volume group "vgSystem" using metadata type lvm2

May 17 14:33:01.997 [INFO] [VolumeManager] [checkVGErrors] [26576579-eb79-4465-af76-b170909b1b14] No IO errors detected for volume group 26576579-eb79-4465-af76-b170909b1b14
```
